### PR TITLE
[SCHEMA] Reorganize schema rendering code and add column table function

### LINKED ISSUE #1286: Create column table-generating macro based on new rule format
We have the metadata table definitions and associated macro working, so we just need the same for columns.

### PR DESCRIPTION
**Pull request by tsalo**
Closes #1286.

To do:

- [x] Remove the old column table function.
- [x] Replace all calls to the old column table function in the spec with the new one.
- [x] Write a test or two.

Changes proposed:

- Split `render.py` into a module with three files: `render/text.py`, `render/tables.py`, and `render/utils.py`.
- Refactor the column table function to use rules.

### CODE REVIEW

**Comment by tsalo** [comment]
How do folks feel about the proposed new columns table format:

image

vs. the old, multi-table format:

image

**Comment by effigies** [comment]
I prefer the new style.

**Comment by Remi-Gau** [comment]
> I prefer the new style.

me too

**Review by effigies — COMMENTED** [review]
Got a little carried away with things that are not entirely in scope of this PR. Feel free to ignore and resolve, but I might use them as reminders to revisit some of this code in the future.

File: src/schema/rules/tabular_data/modality_agnostic.yaml [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+Samples:
+  selectors:
+    - suffix == "samples"
+    - extension == ".tsv"
+  columns:
+    sample_id:
+      level: required
+      description_addendum: |
+        The combination of `sample_id` and `participant_id` MUST be unique.
```
[the remark this answers is not in the document]
Comment (effigies) [reply src/schema/rules/tabular_data/modality_agnostic.yaml:29]: cc @rwblair for thoughts on this one.

File: src/schema/rules/tabular_data/modality_agnostic.yaml [hunk 2]
```diff
@@ -0,0 +1,66 @@
+---
+Participants:
+  selectors:
+    - suffix == "participants"
+    - extension == ".tsv"
```
[the remark this answers is not in the document]
Comment (tsalo) [reply src/schema/rules/tabular_data/modality_agnostic.yaml:5]: Yes, you're right. I only put that in as a placeholder and then completely forgot about it. Thanks for catching it.

File: tools/schemacode/bidsschematools/render/utils.py [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    to avoid adding num2words to our requirements.
+    It only works with the first few numbers.
+
+    Parameters
+    ----------
+    integer : int
+    to : {"ordinal", "cardinal"}, optional
+    """
+    if integer not in [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]:
+        raise ValueError(f"Input {integer} is not supported.")
```
[the remark this answers is not in the document]
Comment (tsalo) [reply tools/schemacode/bidsschematools/render/utils.py]: Thanks! Done in [97361c6](https://github.com/bids-standard/bids-specification/pull/1290/commits/97361c68520397fd6a09c0358d74b7aabf32f28c).

File: tools/schemacode/bidsschematools/render/utils.py [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+        # Use dictionary to get unique substrings while preserving insertion order
+        substrings = {resolve_metadata_type(subdict): None for subdict in definition["anyOf"]}
+
+        string = " or ".join(substrings)
+
+    else:
+        # A hack to deal with $ref in the current schema
+        print(f"Type could not be inferred for {definition['name']}")
+        pprint(definition)
+        string = "unknown"
```
[the remark this answers is not in the document]
Comment (tsalo) [reply tools/schemacode/bidsschematools/render/utils.py]: Thanks! Done in 5b58e8c.

File: src/schema/rules/tabular_data/modality_agnostic.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rwblair) [reply src/schema/rules/tabular_data/modality_agnostic.yaml:27]: With #1112 I don't think we have the tools to do this. I may not be clever enough at the moment.  We can't get indices of values in arrays, we can't iterate over arrays, can't zip or reduce, etc.

My more general understanding of the problem: This is an assertion between two values at the same arbitrary index in two arrays and the rest of the values of those two arrays.

I haven't been able to come up with a single addition to the expression language that would be able to do this.

**Review by effigies — APPROVED** [review]

File: src/schema/rules/tabular_data/modality_agnostic.yaml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  initial_columns:
+    - filename
+  columns:
+    filename:
+      level: required
+      description_addendum: |
+        There MUST be exactly one row for each file.
+    acq_time__scans: optional
+  additional_columns: allowed_if_defined
```
Comment (effigies) [inline src/schema/rules/tabular_data/modality_agnostic.yaml:49]: My reading is that they are permitted in the usual way. In general TSV files RECOMMEND adding definitions for new columns, so this is not a departure:

> All such included additional fields SHOULD be documented in an accompanying _scans.json file that describes these fields in detail (see Tabular files).

```suggestion
  additional_columns: allowed
```

File: src/schema/rules/tabular_data/modality_agnostic.yaml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  initial_columns:
+    - session_id
+  columns:
+    session_id:
+      level: required
+      description_addendum: |
+        There MUST be exactly one row for each session.
+    acq_time__sessions: optional
+    pathology: recommended
+  additional_columns: allowed_if_defined
```
Comment (effigies) [inline src/schema/rules/tabular_data/modality_agnostic.yaml:64]: And again.

```suggestion
  additional_columns: allowed
```

File: src/04-modality-specific-files/04-intracranial-electroencephalography.md [hunk 7]
```diff
@@ -221,45 +221,13 @@ Note that electrode positions SHOULD NOT be added to this file but to `*_electro
...
-  src/schema/objects/columns.yaml
+  src/schema/rules/tabular_data/*.yaml
 and a guide for using macros can be found at
  https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
 -->
-{{ MACROS___make_columns_table(
-   {
-      "name__channels": ("REQUIRED", "When a corresponding electrode is specified in `_electrodes.tsv`, "
-                         "the name of that electrode MAY be specified here and the reference electrode "
-                         "name MAY be provided in the `reference` column."),
```
Comment (effigies) [inline src/04-modality-specific-files/04-intracranial-electroencephalography.md:236]: This description addendum is missing from `rules.tabular_data.ieeg.iEEGChannels`.

Comment (tsalo) [reply src/04-modality-specific-files/04-intracranial-electroencephalography.md:236]: It looks like you handled this so I'll resolve it.

**Comment by effigies** [comment]
Since this has a rendering effect, I would appreciate a review by another maintainer.

**Comment by bendhouseart** [comment]
Taylor, ignore comments leading w/ [note].

File: src/03-modality-agnostic-files.md [hunk 8]
```diff
@@ -224,21 +224,11 @@ following values for them:
 
 <!-- This block generates a columns table.
 The definitions of these fields can be found in
-  src/schema/objects/columns.yaml
+  src/schema/rules/tabular_data/*.yaml
```
Comment (bendhouseart) [inline src/03-modality-agnostic-files.md:227]: [note] checked 👍

File: src/03-modality-agnostic-files.md [hunk 9]
```diff
@@ -224,21 +224,11 @@ following values for them:
...
-      "participant_id": ("REQUIRED", "There MUST be exactly one row for each participant."),
-      "species": "RECOMMENDED",
-      "age": "RECOMMENDED",
-      "sex": "RECOMMENDED",
-      "handedness": "RECOMMENDED",
-      "strain": "RECOMMENDED",
-      "strain_rrid": "RECOMMENDED",
-   }
-) }}
+{{ MACROS___make_columns_table("modality_agnostic.Participants") }}
```
Comment (bendhouseart) [inline src/03-modality-agnostic-files.md:231]: [note] 👍

File: src/03-modality-agnostic-files.md [hunk 10]
```diff
@@ -309,19 +299,11 @@ Each sample MUST be described by one and only one row.
...
-{{ MACROS___make_columns_table(
-   {
-      "sample_id": ("REQUIRED", "The combination of `sample_id` and `participant_id` MUST be unique."),
-      "participant_id": ("REQUIRED", "The combination of `sample_id` and `participant_id` MUST be unique."),
-      "sample_type": "REQUIRED",
-      "pathology": "RECOMMENDED",
-      "derived_from": "RECOMMENDED",
-   }
-) }}
+{{ MACROS___make_columns_table("modality_agnostic.Samples") }}
```
Comment (bendhouseart) [inline src/03-modality-agnostic-files.md:306]: [note] 👍

File: src/03-modality-agnostic-files.md [hunk 11]
```diff
@@ -475,16 +457,11 @@ Such recordings MUST be documented with one row per file
...
 and a guide for using macros can be found at
  https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
 -->
-{{ MACROS___make_columns_table(
-   {
-      "filename": ("REQUIRED", "There MUST be exactly one row for each file."),
-      "acq_time__scans": ("OPTIONAL"),
-   }
-) }}
+{{ MACROS___make_columns_table("modality_agnostic.Scans") }}
```
Comment (bendhouseart) [inline src/03-modality-agnostic-files.md:464]: [note] 👍

File: src/03-modality-agnostic-files.md [hunk 12]
```diff
@@ -524,17 +501,11 @@ Column names in `sessions.tsv` files MUST be different from group level particip
...
  https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
 -->
-{{ MACROS___make_columns_table(
-   {
-      "session_id": ("REQUIRED", "There MUST be exactly one row for each session."),
-      "acq_time__sessions": ("OPTIONAL"),
-      "pathology": "RECOMMENDED",
-   }
-) }}
+{{ MACROS___make_columns_table("modality_agnostic.Sessions") }}
```
Comment (bendhouseart) [inline src/03-modality-agnostic-files.md:508]: [note] 👍

File: src/04-modality-specific-files/02-magnetoencephalography.md [hunk 13]
```diff
@@ -217,42 +217,13 @@ Missing values MUST be indicated with `"n/a"`.
...
-      "sampling_frequency": "OPTIONAL",
-      "low_cutoff": "OPTIONAL",
-      "high_cutoff": "OPTIONAL",
-      "notch": "OPTIONAL",
-      "software_filters": "OPTIONAL",
-      "status": "OPTIONAL",
-      "status_description": "OPTIONAL",
-   }
-) }}
+{{ MACROS___make_columns_table("meg.MEGChannels") }}
```
Comment (bendhouseart) [inline src/04-modality-specific-files/02-magnetoencephalography.md:226]: [note] 👍

File: src/04-modality-specific-files/03-electroencephalography.md [hunk 14]
```diff
@@ -205,42 +205,13 @@ See the examples for `*_channels.tsv` below, and for `*_electrodes.tsv` in
...
-      "sampling_frequency": "OPTIONAL",
-      "reference__eeg": "OPTIONAL",
-      "low_cutoff": "OPTIONAL",
-      "high_cutoff": "OPTIONAL",
-      "notch": "OPTIONAL",
-      "status": "OPTIONAL",
-      "status_description": "OPTIONAL",
-   }
-) }}
+{{ MACROS___make_columns_table("eeg.EEGChannels") }}
```
Comment (bendhouseart) [inline src/04-modality-specific-files/03-electroencephalography.md:214]: type in eeg.EEGChannels does not show the following in Description column, current spec renders below:

```Must be one of: "MEGMAG", "MEGGRADAXIAL", "MEGGRADPLANAR", "MEGREFMAG", "MEGREFGRADAXIAL", "MEGREFGRADPLANAR", "MEGOTHER", "EEG", "ECOG", "SEEG", "DBS", "VEOG", "HEOG", "EOG", "ECG", "EMG", "TRIG", "AUDIO", "PD", "EYEGAZE", "PUPIL", "MISC", "SYSCLOCK", "ADC", "DAC", "HLU", "FITERR", "OTHER".
```

File: src/04-modality-specific-files/03-electroencephalography.md [hunk 15]
```diff
@@ -300,38 +271,13 @@ expected in cartesian coordinates according to the `EEGCoordinateSystem` and
...
  https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
 -->
-{{ MACROS___make_columns_table(
-   {
-      "type__electrodes": "RECOMMENDED",
-      "material": "RECOMMENDED",
-      "impedance": "RECOMMENDED",
-   }
-) }}
+{{ MACROS___make_columns_table("eeg.EEGElectrodes") }}
```
Comment (bendhouseart) [inline src/04-modality-specific-files/03-electroencephalography.md:280]: [note] 👍

File: src/04-modality-specific-files/04-intracranial-electroencephalography.md [hunk 16]
```diff
@@ -221,45 +221,13 @@ Note that electrode positions SHOULD NOT be added to this file but to `*_electro
...
-      "reference__ieeg": "OPTIONAL",
-      "group__channel": ("OPTIONAL", "Note that any groups specified in `_electrodes.tsv` must match those present here."),
-      "sampling_frequency": "OPTIONAL",
-      "description": "OPTIONAL",
-      "notch": "OPTIONAL",
-      "status": "OPTIONAL",
-      "status_description": "OPTIONAL",
-   }
-) }}
+{{ MACROS___make_columns_table("ieeg.iEEGChannels") }}
```
Comment (bendhouseart) [inline src/04-modality-specific-files/04-intracranial-electroencephalography.md:230]: Same issue as eeg type render above, the following are present in current spec, but not this render:
```
Must be one of: "MEGMAG", "MEGGRADAXIAL", "MEGGRADPLANAR", "MEGREFMAG", "MEGREFGRADAXIAL", "MEGREFGRADPLANAR", "MEGOTHER", "EEG", "ECOG", "SEEG", "DBS", "VEOG", "HEOG", "EOG", "ECG", "EMG", "TRIG", "AUDIO", "PD", "EYEGAZE", "PUPIL", "MISC", "SYSCLOCK", "ADC", "DAC", "HLU", "FITERR", "OTHER".
```

Comment (bendhouseart) [reply src/04-modality-specific-files/04-intracranial-electroencephalography.md:230]: See associated glossary entry @bendhouseart

File: src/04-modality-specific-files/03-electroencephalography.md (hunk 14 above)
Comment (bendhouseart) [reply src/04-modality-specific-files/03-electroencephalography.md:214]: See associated glossary entry @bendhouseart

Comment (bendhouseart) [reply src/04-modality-specific-files/03-electroencephalography.md:214]: [note] 👍

File: src/04-modality-specific-files/04-intracranial-electroencephalography.md [hunk 17]
```diff
@@ -360,56 +328,13 @@ The order of the required columns in the `*_electrodes.tsv` file MUST be as list
...
- https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
--->
-{{ MACROS___make_columns_table(
-   {
-      "type__electrodes": "OPTIONAL",
-      "impedance": "OPTIONAL",
-      "dimension": "OPTIONAL",
-   }
-) }}
+{{ MACROS___make_columns_table("ieeg.iEEGElectrodes") }}
```
Comment (bendhouseart) [inline src/04-modality-specific-files/04-intracranial-electroencephalography.md:337]: [note] 👍

File: src/04-modality-specific-files/05-task-events.md [hunk 18]
```diff
@@ -36,21 +36,11 @@ and OPTIONAL columns:
...
-      "onset": "REQUIRED",
-      "duration": "REQUIRED",
-      "sample": "OPTIONAL",
-      "trial_type": "OPTIONAL",
-      "response_time": "OPTIONAL",
-      "value": "OPTIONAL",
-      "HED": "OPTIONAL",
-   }
-) }}
+{{ MACROS___make_columns_table("task.TaskEvents") }}
```
Comment (bendhouseart) [inline src/04-modality-specific-files/05-task-events.md:43]: [note] 👍

File: src/04-modality-specific-files/06-physiological-and-other-continuous-recordings.md [hunk 19]
```diff
@@ -149,17 +149,11 @@ following naming conventions SHOULD be used for the column names:
...
  https://github.com/bids-standard/bids-specification/blob/master/macros_doc.md
 -->
-{{ MACROS___make_columns_table(
-   {
-      "cardiac": "OPTIONAL",
-      "respiratory": "OPTIONAL",
-      "trigger": "OPTIONAL",
-   }
-) }}
+{{ MACROS___make_columns_table("physio.PhysioColumns") }}
```
Comment (bendhouseart) [inline src/04-modality-specific-files/06-physiological-and-other-continuous-recordings.md:156]: [note] 👍

File: src/04-modality-specific-files/09-positron-emission-tomography.md [hunk 20]
```diff
@@ -347,20 +347,11 @@ The `time` column MUST always be the first column.
...
-   {
-      "time": "REQUIRED",
-      "plasma_radioactivity": "REQUIRED if `PlasmaAvail` is `true`",
-      "metabolite_parent_fraction": "REQUIRED if `MetaboliteAvail` is `true`",
-      "metabolite_polar_fraction": "RECOMMENDED if `MetaboliteAvail` is `true`",
-      "hplc_recovery_fractions": "REQUIRED if `MetaboliteRecoveryCorrectionApplied` is `true`",
-      "whole_blood_radioactivity": "REQUIRED if `WholeBloodAvail` is `true`",
-   }
-) }}
+{{ MACROS___make_columns_table("pet.Blood") }}
```
Comment (bendhouseart) [inline src/04-modality-specific-files/09-positron-emission-tomography.md:354]: [note] 👍

File: src/05-derivatives/03-imaging.md [hunk 21]
```diff
@@ -520,19 +520,11 @@ These TSV lookup tables contain the following columns:
...
-{{ MACROS___make_columns_table(
-   {
-      "index": "REQUIRED",
-      "name__segmentations": "REQUIRED",
-      "abbreviation": "OPTIONAL",
-      "color": "OPTIONAL",
-      "mapping": "OPTIONAL",
-   }
-) }}
+{{ MACROS___make_columns_table("derivatives.common_derivatives.SegmentationLookup") }}
```
Comment (bendhouseart) [inline src/05-derivatives/03-imaging.md:527]: [note] 👍

**Review by bendhouseart — APPROVED** [review]
Tables render, many 👍

**Pull request merged**